pkg/termcorder/termcorder.go

```diff
@@ -0,0 +1,210 @@
+		return nil
+	}
+}
+
+//NewTermcording returns a pointer to a new variable of type Termcording given a config
+//variable and (functional) options.
+func NewTermcording(c *Config, options ...func(*Termcording) error) (*Termcording, error) {
+	tc := &Termcording{
+		Config: c,
+	}
+	for _, option := range options {
+		err := option(tc)
+		if err != nil {
+			return &Termcording{}, err
+		}
+	}
+	return tc, nil
+}
+
+//TermcordingFromFlags parses command line flags (and arguments) and returns a pointer to a
+//new variable of type `Termcording`.
+func TermcordingFromFlags(options ...func(*Termcording) error) (*Termcording, error) {
+	var fName, cmdName string
+	var cmdArgs []string
+
+	var h, q, a, k bool
+
+	cli = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	cli.BoolVar(&h, "h", false, "Prints this message")
+	cli.BoolVar(&q, "q", false, "Quiet mode - suppresses the recording start and end prompts")
+	cli.BoolVar(&a, "a", false, "Appends to file instead of overwriting it")
+	cli.BoolVar(&k, "k", false, "Log key strokes to file as well")
+
+	cli.Parse(os.Args[1:])
+
+	shell, ok := os.LookupEnv("SHELL")
+	if cli.Arg(1) == "" && (!ok || shell == "") {
+		return &Termcording{}, errors.New("shell not set")
+	}
+	interactive := true
+	switch cli.NArg() {
+	case 0:
+		fName = defaultFileName
+		cmdName = shell
+	case 1:
+		fName = cli.Arg(0)
+		cmdName = shell
+	default:
+		fName = cli.Arg(0)
+		cmdName = cli.Arg(1)
+		cmdArgs = cli.Args()[2:]
+		interactive = false
+	}
+
+	return NewTermcording(&Config{
+		Filename:      fName,
+		CmdName:       cmdName,
+		CmdArgs:       cmdArgs,
+		QuietMode:     q,
+		Append:        a,
+		Interactive:   interactive,
+		LogKeystrokes: k,
+		PrintHelp:     h,
+	}, options...)
+}
+
+//Start runs the tc.cmd in a pseudo-terminal and writes all output to tc.out
+func (tc *Termcording) Start() error {
+	if tc.Config.PrintHelp {
+		if tc.out == nil {
+			tc.out = os.Stdout
+		}
+		printHelp(tc)
+		return nil
+	}
+
+	if tc.cmd == nil {
+		tc.cmd = exec.Command(tc.Config.CmdName, tc.Config.CmdArgs...)
+	}
+
+	if !tc.Config.QuietMode {
+		fmt.Println("Starting recording session. CTRL-D to end.")
+		defer fmt.Printf("\nRecording session ended. Session saved to %s\n", tc.Config.Filename)
+	}
+
+	ptmx, restoreMode, err := ptmxFromCmd(tc.cmd, tc.Config.Interactive) //TODO convert to method.
+	if err != nil {
+		return err
+	}
+	defer func() {
+		restoreMode()
+		ptmx.Close()
+	}()
+
+	in := io.Writer(ptmx)
+	if tc.out == nil {
+		mode := os.O_TRUNC
+		if tc.Config.Append {
+			mode = os.O_APPEND
+		}
+		f, err := os.OpenFile(tc.Config.Filename, os.O_WRONLY|os.O_CREATE|mode, 0755)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if tc.Config.LogKeystrokes {
+			in = io.MultiWriter(ptmx, f)
+		}
+		tc.out = io.MultiWriter(os.Stdout, f)
+	}
+
+	go func() {
+		io.Copy(in, os.Stdin)
+	}()
+
+	io.Copy(tc.out, ptmx)
+
+	return tc.cmd.Wait()
+}
+
+func ptmxFromCmd(c *exec.Cmd, interactive bool) (*os.File, func(), error) {
```

Comment:
I can guess that the file result is the recording file, but I can't guess what the `func()` is! Naming the results might help here.

Reply:
Actually, the `*os.File` return value here is the pseudo-terminal "file"! So named returns is definitely needed here 😅 
Renaming the function should help too.